### PR TITLE
[53500]feature history screen

### PR DESCRIPTION
**Pull request by ghost**
## Related Tickets
- [#Task 53501: Create UI](https://edu-redmine.sun-asterisk.vn/issues/53501)
- [#Task 53502: Handler event](https://edu-redmine.sun-asterisk.vn/issues/53502)
- [#Task 53503:get data](https://edu-redmine.sun-asterisk.vn/issues/53503)
## WHAT
- 
## Evidence (Screenshot or Video)
image

## Review Checklist

Category | View Point | Description | Self review | Reviewer2 (name)
--- | --- | --- | --- | ---
Conventions | Sun* coding conventions | https://sal.vn/WtYPAy |<li>- [ ] yes</li>|<li>- [ ] yes</li>
Conventions | Kotlin coding conventions | https://kotlinlang.org/docs/coding-conventions.html |<li>- [ ] yes</li>|<li>- [ ] yes</li>
Redmine | Sun* Redmine working process  | https://sal.vn/piPoQn |<li>- [ ] yes</li>|<li>- [ ] yes</li>

## Notes (Optional)
*(Impacted Areas in Application(List features, api, models or services that this PR will affect))*

*(List gem, library third party add new)*

*(Other notes)*

### CODE REVIEW

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -45,6 +54,13 @@
...
+            android:name=".screen.FcmService"
+            android:exported="false">
+            <intent-filter>
+                <action android:name="com.google.firebase.INSTANCE_ID_EVENT"/>
+            </intent-filter>
+        </service>
     </application>
 
-</manifest>
+</manifest>
```
Comment (minhnn-1279) [inline app/src/main/AndroidManifest.xml]: endline

File: app/src/main/java/com/example/finalapplication/screen/MyApplication.kt [hunk 2]
```diff
@@ -26,7 +42,38 @@ val appModule = module {
     viewModel {
         ProfileViewModel(get())
     }
-    single<UserDataSource.remote> { RemoteUser() }
+    viewModel {
+        SearchViewModel(get())
+    }
+    fun provideRetrofit(factory: Gson) = Retrofit.Builder()
+        .baseUrl("https://fcm.googleapis.com/fcm/")
```
Comment (minhnn-1279) [inline app/src/main/java/com/example/finalapplication/screen/MyApplication.kt]: tách ra string BASE_URL

File: app/src/main/java/com/example/finalapplication/screen/chatroom/ChatRoomActivity.kt [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    override fun initData() {
+        val id = intent.getStringExtra(Constant.RECIVER)
+        chatViewModel.getReciver(id)
+        // chatViewModel.getNewMessage(currentUser.id+id)
+
+    }
+
+    override fun dispatchTouchEvent(ev: MotionEvent?): Boolean {
+        if (currentFocus != null ) {
+            val imm = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
```
Comment (minhnn-1279) [inline app/src/main/java/com/example/finalapplication/screen/chatroom/ChatRoomActivity.kt]: tên biến clear hơn nhé a :v

File: app/src/main/java/com/example/finalapplication/screen/chatroom/ChatViewModel.kt [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+
+            })
+        }, isShowLoading = false)
+    }
+
+    fun getReciver(id: String?) {
+        launchTask<User>(onRequest = {
+            userRepository.getUser(id, object : Listenner<User> {
+                override fun onSuccess(data: User) {
+                    _reciver.value = data
```
Comment (minhnn-1279) [inline app/src/main/java/com/example/finalapplication/screen/chatroom/ChatViewModel.kt]: sai trính tả :v

File: app/src/main/res/layout/activity_incoming.xml [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+            <ImageView
+                android:id="@+id/button_deny"
+                android:layout_width="match_parent"
+                android:layout_height="match_parent"
+                android:layout_margin="@dimen/dp_10"
+                android:src="@drawable/ic_baseline_clear_24" />
+
+        </androidx.cardview.widget.CardView>
+
+</androidx.constraintlayout.widget.ConstraintLayout>
```
Comment (minhnn-1279) [inline app/src/main/res/layout/activity_incoming.xml]: .

File: app/src/main/res/layout/activity_outgoing.xml [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    android:layout_height="match_parent"
+    tools:context=".screen.outgoing.OutgoingActivity">
+
+    <TextView
+        android:id="@+id/text_call_type"
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:layout_marginTop="@dimen/dp_60"
+        android:gravity="center"
+        android:text="Cuộc gọi video đến..."
```
Comment (minhnn-1279) [inline app/src/main/res/layout/activity_outgoing.xml]: hardcode text nhé a

**Review by ghost — APPROVED** [review]

**Review by minhnn-1279 — APPROVED** [review]

File: app/src/main/java/com/example/finalapplication/screen/chatroom/ChatRoomActivity.kt [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+        val id = intent.getStringExtra(Constant.RECIVER)
+        chatViewModel.getReciver(id)
+        // chatViewModel.getNewMessage(currentUser.id+id)
+
+    }
+
+    override fun dispatchTouchEvent(ev: MotionEvent?): Boolean {
+        if (currentFocus != null ) {
+            val inputMethodManager = getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
+            inputMethodManager.hideSoftInputFromWindow(currentFocus!!.windowToken, 0)
```
Comment (vietnh-1109) [inline app/src/main/java/com/example/finalapplication/screen/chatroom/ChatRoomActivity.kt]: ẩn , hiện bàn phím thì cho ra extension hoặc base activity

File: app/src/main/java/com/example/finalapplication/screen/chatroom/ChatViewModel.kt [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+import com.example.finalapplication.data.repository.MessageRepositoryImpl
+import com.example.finalapplication.data.repository.UserRepositoryIpml
+import com.example.finalapplication.data.repository.resource.Listenner
+import com.example.finalapplication.utils.Constant
+import com.example.finalapplication.utils.base.BaseViewModel
+
+class ChatViewModel(
+    private val userRepository: UserRepositoryIpml,
+    private val messageRepository: MessageRepositoryImpl,
+    private val contactRepository: ContactRepositoryImpl
```
Comment (vietnh-1109) [inline app/src/main/java/com/example/finalapplication/screen/chatroom/ChatViewModel.kt]: truyền interface chứ nhỉ ?

File: app/src/main/java/com/example/finalapplication/screen/chatroom/ChatViewModel.kt [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+
+    fun getNewMessage(id: String) {
+        launchTask<Contact>(onRequest = {
+            contactRepository.getContact(id, object : Listenner<Contact> {
+                override fun onSuccess(data: Contact) {
+                    _newMessage.value = data.message
+                }
+
+                override fun onError(msg: String) {
+                    _newMessage.value = null
```
Comment (vietnh-1109) [inline app/src/main/java/com/example/finalapplication/screen/chatroom/ChatViewModel.kt]: bỏ nhé

File: app/src/main/java/com/example/finalapplication/screen/chatroom/ChatViewModel.kt [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+                reciverId,
+                lastIndex,
+                object : Listenner<List<Message>> {
+                    override fun onSuccess(data: List<Message>) {
+                        _historyMessage.value = data
+                    }
+
+                    override fun onError(msg: String) {
+                        message.value = (Constant.ERROR_)
+                        _historyMessage.value = null
```
Comment (vietnh-1109) [inline app/src/main/java/com/example/finalapplication/screen/chatroom/ChatViewModel.kt]: tương tự

File: app/src/main/java/com/example/finalapplication/screen/historycontact/HistoryContactListAdapter.kt [hunk 11]
```diff
@@ -0,0 +1,90 @@
...
+                    textMessage.setTextColor(Color.BLACK)
+                    textMessage.setTypeface(null, Typeface.BOLD)
+                    textTime.setTypeface(null, Typeface.BOLD)
+                }
+                textUsername.text = contact.adversaryUser?.name
+                textTime.text = contact.message?.time?.toDateString().toString()
+                when(contact.message?.chatType){
+                    Message.chatbox -> {
+                        if(contact.message?.text.isNullOrEmpty()) textMessage.text = Message.image
+                            else textMessage.text = contact.message?.text
```
Comment (vietnh-1109) [inline app/src/main/java/com/example/finalapplication/screen/historycontact/HistoryContactListAdapter.kt]:
```suggestion
                        textMessage = if(contact.message?.text.isNullOrEmpty())  Message.image
                            else contact.message?.text
```

File: app/src/main/java/com/example/finalapplication/screen/historycontact/HistoryContactViewModel.kt [hunk 12]
```diff
@@ -0,0 +1,37 @@
+package com.example.finalapplication.screen.historycontact
+
+import android.util.Log
```
Comment (vietnh-1109) [inline app/src/main/java/com/example/finalapplication/screen/historycontact/HistoryContactViewModel.kt]: .

File: app/src/main/java/com/example/finalapplication/screen/historycontact/HistoryContactViewModel.kt [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+import androidx.lifecycle.LiveData
+import androidx.lifecycle.MutableLiveData
+import com.example.finalapplication.data.model.Contact
+import com.example.finalapplication.data.repository.ContactRepositoryImpl
+import com.example.finalapplication.data.repository.resource.Listenner
+import com.example.finalapplication.utils.base.BaseViewModel
+
+class HistoryContactViewModel(private val contactRepository: ContactRepositoryImpl) :
```
Comment (vietnh-1109) [inline app/src/main/java/com/example/finalapplication/screen/historycontact/HistoryContactViewModel.kt]: truyền interface

**Pull request merged**